compiler/src/iree/compiler/Codegen/LLVMGPU/LLVMGPUVectorLowering.cpp

```diff
@@ -183,6 +184,321 @@ struct SetMulAddFMF final : OpRewritePattern<vector::MultiDimReductionOp> {
+        return i;
+    }
+    return std::nullopt;
+  }
+
+  static SmallVector<int64_t> getReductionIndex(AffineMap map,
+                                                ArrayAttr iteratorTypes) {
+    SmallVector<int64_t> dimsIdx;
+    for (unsigned i = 0, e = map.getNumResults(); i < e; i++) {
+      if (vector::isReductionIterator(iteratorTypes[map.getDimPosition(i)]))
+        dimsIdx.push_back(i);
+    }
+    return dimsIdx;
+  }
+
+  static SmallVector<int64_t> invert(ArrayRef<int64_t> perm) {
+    SmallVector<int64_t> inv(perm.size());
+    for (auto [i, p] : llvm::enumerate(perm)) {
+      inv[p] = i;
+    }
+    return inv;
+  }
+
+  static void getReductionAndParallelLoopDims(ArrayAttr iters,
+                                              SmallVectorImpl<int64_t> &red,
+                                              SmallVectorImpl<int64_t> &par) {
+    for (auto [idx, attr] : llvm::enumerate(iters)) {
+      if (vector::isReductionIterator(attr)) {
+        red.push_back(idx);
+      } else {
+        par.push_back(idx);
+      }
+    }
+  }
+
+  /// Constructs a permutation for vector.transpose from an affine map and a
+  /// reordered list of dimension.
+  ///
+  /// Example:
+  ///   map: (d0, d1, d2) -> (d0, d2, d1)
+  ///   iterator_types = ["parallel","parallel","reduction"]
+  //    ==> new dim order: [2, 0, 1]
+  ///
+  ///   Step 1: Build dim-to-result mapping from the map.
+  ///           dimToRes = [0, 2, 1] i.e {0: 0, 1: 2, 2: 1}
+  ///
+  ///   Step 2: Walk new dimension order in order to build permutation.
+  ///           indices[0]=2 -> dimToRes[2]=1
+  ///           indices[1]=0 -> dimToRes[0]=0
+  ///           indices[2]=1 -> dimToRes[1]=2
+  ///
+  ///   Result: perm = [1, 0, 2]
+  static SmallVector<int64_t>
+  getPermutationFromIndexingMap(AffineMap map, ArrayRef<int64_t> indices) {
+    SmallVector<int64_t> dimToRes(map.getNumDims());
+    for (int res = 0, e = map.getNumResults(); res != e; ++res) {
+      dimToRes[map.getDimPosition(res)] = res;
+    }
+
+    return to_vector(
+        llvm::map_range(indices, [&](int64_t i) { return dimToRes[i]; }));
+  }
+
+  static int64_t productOfDims(VectorType vt, unsigned lo, unsigned hi) {
+    int64_t p = 1;
+    for (unsigned i = lo; i < hi; ++i) {
+      p *= vt.getDimSize(i);
+    }
+    return p;
+  }
+
+  static bool isIdentityPermutation(ArrayRef<int64_t> perm) {
+    return llvm::all_of(llvm::enumerate(perm),
+                        [](auto p) { return p.value() == p.index(); });
+  }
+
+  static Value buildFMAChain(PatternRewriter &rewriter, Location loc,
+                             Value lhs2D, Value rhs2D, Value accFlat,
+                             int64_t K) {
+    Value current = accFlat;
+
+    for (int64_t k = K - 1; k >= 0; --k) {
+      Value a = vector::ExtractOp::create(rewriter, loc, lhs2D, k);
+      Value b = vector::ExtractOp::create(rewriter, loc, rhs2D, k);
+      current = math::FmaOp::create(rewriter, loc, a, b, current);
```

Comment:
If `math.fma` can't handle vector types, it should be able to

Also, if this _does_ need to be extracts, I'd like to flag `StaticTileOffsetRange` so that you don't need to flatten everything

Reply:
... Ok, yeah, math.fma supports vectors, and that'll let you get to LLVM vectors (I hope), which'll let you get packed FMAs where they exist

Reply:
Ah, this is an extract of a 1-D vector, not elementwise. Fair enough